torch/distributed/_shard/replicated_tensor.py

```diff
@@ -0,0 +1,118 @@
+
+    NOTE: We do not gurantee equal content of ReplicatedTensor across nodes after its
+    construction. Although we defined proper inter-op rules to make sure ReplicatedTensor
+    stays the same, there's no enforcement on it (i.e. if you manually modify content on
+    some ranks, the modified value will not automatically get synced to other nodes). If
+    you wish to manually validate tensors are the same across ranks, use `validate()`.
+
+    """
+    def __new__(cls, data=None):
+        if data is None:
+            data = torch.empty(0)
+        return torch.Tensor._make_subclass(cls, data)
+
+    def __repr__(self):
+        return f"ReplicatedTensor({super(ReplicatedTensor, self).__repr__()})"
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        new_args = []
+        new_kwargs = {}
+
+        # check if args/kwargs have tensor (non-replicated) operands, we have to do
+        # this so that we can detect there's a torch.Tensor operand and thus not
+        # converting results back to ReplicatedTensor
+        has_tensor = False
+        for arg in args:
+            if isinstance(arg, torch.Tensor) and not isinstance(arg, ReplicatedTensor):
+                has_tensor = True
+
+        if kwargs is not None:
+            for k, v in kwargs.items():
+                if isinstance(v, torch.Tensor) and not isinstance(v, ReplicatedTensor):
+                    has_tensor = True
+
+        # do dispatch base on the inter-op rules we defined
+        for arg in args:
+            if isinstance(arg, ShardedTensor):
+                # redispatch to ShardedTensor
+                # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                return arg.__torch_function__(func, types, args, kwargs)
+            else:
+                new_args.append(arg)
+
+
+        if kwargs is not None:
+            for k, v in kwargs.items():
+                if isinstance(v, ShardedTensor):
+                    # redispatch to ShardedTensor
+                    # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                    return v.__torch_function__(func, types, args, kwargs)
+                else:
+                    new_kwargs[k] = v
+
+        # We cann't do super().__torch_function__() as it implicitly convert the result
+        # back to tensor subclasses, where in our case, we need to control the output type
+        # base on the inter-op rules we defined.
+        with torch._C.DisableTorchFunction():
+            rs = func(*new_args, **new_kwargs)
+            if func in get_default_nowrap_functions():
+                return rs
+            if not has_tensor and isinstance(rs, torch.Tensor) and not isinstance(rs, cls):
+                # if it does not have tensor operands and does not get dispatched to ShardedTensor
+                # __torch_function__, we return a ReplicatedTensor according to our inter-op rule
+                rs = rs.as_subclass(cls)
```

Comment:
Does this cover all cases? What happens if we add a scalar to ReplicatedTensor (which is different on each rank)? Doesn't this check pass and we incorrectly create a ReplicatedTensor?

Reply:
Yeah agreed, should we just enforce the rule that Anything else inter-op with ReplicatedTensor, will result in the type not being ReplicatedTensor anymore? That might make our life easier as we cannot ensure the operand to be the same unless it marked as replicated.

Reply:
> Anything else inter-op with ReplicatedTensor, will result in the type not being ReplicatedTensor anymore?

Yes, I think that is a safer rule.